pkg/cloudprovider/vm.go

```diff
@@ -168,16 +169,39 @@ func (i *vmIdentityInfo) SetUserIdentities(ids map[string]bool) bool {
 
 	// all identities are the node are to be removed
 	if len(nodeList) == 0 {
-		i.info.UserAssignedIdentities = nil
-		if i.info.Type == compute.ResourceIdentityTypeSystemAssignedUserAssigned {
-			i.info.Type = compute.ResourceIdentityTypeSystemAssigned
-		} else {
-			i.info.Type = compute.ResourceIdentityTypeNone
-		}
+		i.ResetResourceIdentityType()
 		return true
 	}
 
 	i.info.Type = getUpdatedResourceIdentityType(i.info.Type)
 	i.info.UserAssignedIdentities = userAssignedIdentities
 	return len(i.info.UserAssignedIdentities) > 0
 }
+
+func (i *vmIdentityInfo) RemoveUserIdentity(delID string) bool {
+	removed := false
+	if i.info.UserAssignedIdentities != nil {
+		for id := range i.info.UserAssignedIdentities {
+			if strings.EqualFold(id, delID) {
+				delete(i.info.UserAssignedIdentities, id)
+				removed = true
+				break
+			}
+		}
+	}
+
+	// Modify identity type in case we remove the last identity from the VM
+	if len(i.info.UserAssignedIdentities) == 0 {
```

Comment:
`RemoveUserIdentity` is removing the identities from the desired list we are going to set (list which contains identities we want to assign and un-assign). So at the end of this, if `len(i.info.UserAssignedIdentities) == 0` then we should not be resetting the resource identity type. Length of 0 indicates we have no operation to perform, right?  But the identities which are still on the node should not be cleaned up.

Reply:
Makes sense. Let me modify the logic.